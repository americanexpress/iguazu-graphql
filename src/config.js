/*
 * Copyright 2018 American Express
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const endpoints = new Map();

// function addEndpoint({ name, ...opts }) {
function addEndpoint(name, opts) {
  // const { name } = opts;
  if (!name || typeof name !== 'string') {
    throw new Error(`name must be a string (was ${name ? typeof name : 'falsey'})`);
  }

  if (endpoints.has(name)) {
    throw new Error(`cannot override existing config for ${name}`);
  }

  endpoints.set(name, opts);
}

function getEndpoint(name) {
  if (!endpoints.has(name)) {
    throw new Error(`${name} is not a configured endpoint`);
  }
  // return {
  //   ...endpoints.get(name),
  //   name,
  // };
  return endpoints.get(name);
}

const config = {
  baseFetch: fetch,
  // don't like the name, but it's part of iguazu-rest so it's consistent ðŸ˜¬
  getToState: state => state.iguazuGraphQL,
  addEndpoint,
  getEndpoint,
};

export function configureIguazuGraphQL(customConfig) {
  Object.assign(config, customConfig, { addEndpoint, getEndpoint });
}

export default config;
