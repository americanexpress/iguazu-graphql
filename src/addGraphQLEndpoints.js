/*
 * Copyright 2018 American Express
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import config from './config';

function addGraphQLEndpoint({ name, fetch, opts }) {
  if (!name || typeof name !== 'string') {
    throw new Error(`required param name must be a string (was ${name ? typeof name : 'falsey'})`);
  }

  if (!fetch || (typeof fetch !== 'function')) {
    throw new Error(`required param fetch must be a function (was ${typeof fetch})`);
  }

  if (opts && typeof opts !== 'object') {
    throw new Error(`param opts must be an object (was ${typeof opts})`);
  }

  config.addEndpoint(name, { fetch, opts });
}

export default function addGraphQLEndpoints(arg) {
  const endpoints = Array.isArray(arg) ? arg : [arg];
  endpoints.forEach((endpoint) => addGraphQLEndpoint(endpoint));
}
