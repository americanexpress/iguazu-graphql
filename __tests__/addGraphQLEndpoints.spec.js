/*
 * Copyright 2018 American Express
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import 'fetch-everywhere';
import config from '../src/config';

import addGraphQLEndpoints from '../src/addGraphQLEndpoints';

describe('addGraphQLEndpoints', () => {
  it('is a function', () => expect(addGraphQLEndpoints).toBeInstanceOf(Function));

  it('throws when not given a name for the endpoint', () => {
    expect(() => addGraphQLEndpoints({})).toThrowErrorMatchingSnapshot();
  });

  it('throws when not given a name as a String for the endpoint', () => {
    expect(() => addGraphQLEndpoints({ name: 9 })).toThrowErrorMatchingSnapshot();
  });

  it('throws when not given a fetch option', () => {
    expect(() => addGraphQLEndpoints({ name: 'throws-when-not-given-a-fetch-option' })).toThrowErrorMatchingSnapshot();
  });

  it('throws when not given a Function for the fetch option', () => {
    expect(() => addGraphQLEndpoints({
      name: 'throws-when-not-given-a-Function-for-the-fetch-option',
      fetch: true,
    })).toThrowErrorMatchingSnapshot();
  });

  it('accepts a string as the name option', () => {
    expect(() => addGraphQLEndpoints({
      name: 'accepts-a-string-as-the-name-option',
      fetch: jest.fn(),
    })).not.toThrowError();
  });

  it('accepts a function as the fetch option', () => {
    expect(() => addGraphQLEndpoints({
      name: 'accepts-a-function-as-the-fetch-option',
      fetch: jest.fn(),
    })).not.toThrowError();
  });

  it('does not throw when not given an opts option', () => {
    expect(() => addGraphQLEndpoints({
      name: 'does-not-throw-when-not-given-an-opts-option',
      fetch: jest.fn(),
    })).not.toThrowError();
  });

  it('throws when not given a object as the opts option', () => {
    expect(() => addGraphQLEndpoints({
      name: 'throws-when-not-given-a-object-as-the-opts-option',
      fetch: jest.fn(),
      opts: 11,
    })).toThrowErrorMatchingSnapshot();
  });

  it('accepts an object as the opts option', () => {
    expect(() => addGraphQLEndpoints({
      name: 'accepts-an-object-as-the-opts-option',
      fetch: jest.fn(),
      opts: { headers: { 'X-CSRF': 'totally-unique' } },
    })).not.toThrowError();
  });

  it('stores the endpoint in the config', () => {
    const name = 'stores-the-endpoint-in-the-config';
    const fetch = jest.fn();
    addGraphQLEndpoints({ name, fetch });
    expect(config.getEndpoint(name)).toHaveProperty('fetch', fetch);
  });

  it('accepts an array of endpoints', () => {
    const fetch1 = jest.fn();
    const fetch2 = jest.fn();

    addGraphQLEndpoints([
      { name: 'accepts-an-array-of-endpoints-1', fetch: fetch1 },
      { name: 'accepts-an-array-of-endpoints-2', fetch: fetch2 },
    ]);
    expect(config.getEndpoint('accepts-an-array-of-endpoints-1')).toHaveProperty('fetch', fetch1);
    expect(config.getEndpoint('accepts-an-array-of-endpoints-2')).toHaveProperty('fetch', fetch2);
  });
});
