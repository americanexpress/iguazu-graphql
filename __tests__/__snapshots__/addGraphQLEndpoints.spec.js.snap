// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addGraphQLEndpoints throws when not given a Function for the fetch option 1`] = `"required param fetch must be a function (was boolean)"`;

exports[`addGraphQLEndpoints throws when not given a fetch option 1`] = `"required param fetch must be a function (was undefined)"`;

exports[`addGraphQLEndpoints throws when not given a name as a String for the endpoint 1`] = `"required param name must be a string (was number)"`;

exports[`addGraphQLEndpoints throws when not given a name for the endpoint 1`] = `"required param name must be a string (was falsey)"`;

exports[`addGraphQLEndpoints throws when not given a object as the opts option 1`] = `"param opts must be an object (was number)"`;
